File: packages/go-routerbase/client_test.go
package routerbase

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestChatCompletion(t *testing.T) {
	var captured struct {
		Authorization string
		ContentType   string
		Body          ChatCompletionRequest
	}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/chat/completions" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}

		captured.Authorization = r.Header.Get("Authorization")
		captured.ContentType = r.Header.Get("Content-Type")
		if err := json.NewDecoder(r.Body).Decode(&captured.Body); err != nil {
			t.Fatal(err)
		}

		_, _ = w.Write([]byte(`{"choices":[{"message":{"role":"assistant","content":"Hello"}}]}`))
	}))
	defer server.Close()

	client := NewClient("sk-rb-test", WithBaseURL(server.URL), WithHTTPClient(server.Client()))
	resp, err := client.ChatCompletion(context.Background(), ChatCompletionRequest{
		Messages: []Message{{Role: "user", Content: "Hi"}},
	})
	if err != nil {
		t.Fatal(err)
	}

	if captured.Authorization != "Bearer sk-rb-test" {
		t.Fatalf("unexpected auth header: %s", captured.Authorization)
	}
	if captured.ContentType != "application/json" {
		t.Fatalf("unexpected content type: %s", captured.ContentType)
	}
	if captured.Body.Model != DefaultModel {
		t.Fatalf("unexpected model: %s", captured.Body.Model)
	}
	if got := resp.Choices[0].Message.Content; got != "Hello" {
		t.Fatalf("unexpected response text: %s", got)
	}
}

func TestListModels(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/models" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}

		_, _ = w.Write([]byte(`{"data":[{"id":"google/gemini-2.5-flash"}]}`))
	}))
	defer server.Close()

	client := NewClient("sk-rb-test", WithBaseURL(server.URL), WithHTTPClient(server.Client()))
	resp, err := client.ListModels(context.Background())
	if err != nil {
		t.Fatal(err)
	}

	if got := resp.Data[0].ID; got != "google/gemini-2.5-flash" {
		t.Fatalf("unexpected model id: %s", got)
	}
}

func TestChatCompletionRequiresMessages(t *testing.T) {
	client := NewClient("sk-rb-test")
	_, err := client.ChatCompletion(context.Background(), ChatCompletionRequest{})
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestAPIError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		_, _ = w.Write([]byte(`{"error":{"message":"bad request"}}`))
	}))
	defer server.Close()

	client := NewClient("sk-rb-test", WithBaseURL(server.URL), WithHTTPClient(server.Client()))
	_, err := client.ListModels(context.Background())
	if err == nil {
		t.Fatal("expected error")
	}
}

File: packages/ruby-routerbase/test/client_test.rb
require "json"
require "minitest/autorun"
require "routerbase"

class FakeTransport
  Response = Struct.new(:code, :body, keyword_init: true)

  attr_reader :requests

  def initialize(response)
    @response = response
    @requests = []
  end

  def request(method:, uri:, headers:, body: nil)
    requests << {
      method: method,
      uri: uri,
      headers: headers,
      body: body
    }
    @response
  end
end

class RouterBaseClientTest < Minitest::Test
  def test_chat_completion
    transport = FakeTransport.new(
      FakeTransport::Response.new(
        code: "200",
        body: JSON.generate(
          "choices" => [
            {
              "message" => {
                "role" => "assistant",
                "content" => "Hello"
              }
            }
          ]
        )
      )
    )
    client = RouterBase::Client.new(api_key: "sk-rb-test", transport: transport)
    response = client.chat_completion(messages: [{ role: "user", content: "Hi" }])
    request = transport.requests.first

    assert_equal "Hello", response.fetch("choices").first.fetch("message").fetch("content")
    assert_equal :post, request.fetch(:method)
    assert_equal "https://routerbase.com/v1/chat/completions", request.fetch(:uri).to_s
    assert_equal "Bearer sk-rb-test", request.fetch(:headers).fetch("Authorization")
    assert_equal RouterBase::DEFAULT_MODEL, JSON.parse(request.fetch(:body)).fetch("model")
  end

  def test_list_models
    transport = FakeTransport.new(
      FakeTransport::Response.new(
        code: "200",
        body: JSON.generate("data" => [{ "id" => "google/gemini-2.5-flash" }])
      )
    )
    client = RouterBase::Client.new(api_key: "sk-rb-test", transport: transport)
    response = client.list_models

    assert_equal "google/gemini-2.5-flash", response.fetch("data").first.fetch("id")
    assert_equal :get, transport.requests.first.fetch(:method)
    assert_equal "https://routerbase.com/v1/models", transport.requests.first.fetch(:uri).to_s
  end

  def test_requires_api_key
    assert_raises(ArgumentError) do
      RouterBase::Client.new(api_key: "")
    end
  end

  def test_api_error
    transport = FakeTransport.new(
      FakeTransport::Response.new(
        code: "400",
        body: JSON.generate("error" => { "message" => "bad request" })
      )
    )
    client = RouterBase::Client.new(api_key: "sk-rb-test", transport: transport)

    error = assert_raises(RouterBase::Error) do
      client.list_models
    end
    assert_match(/bad request/, error.message)
  end
end
